### PR TITLE
138 migration

### PR DESCRIPTION
**Pull request by andrey-katz-systemorph**
Implements migration and provides the explanation of the migration mode.

### CODE REVIEW

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                "## Initialize"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "Note, that when the schema is created, it is not filled autmatically via running the [Configure](../Configure) notebook. Therefore there is a need to initaliaze the DB manually via [this](../../Initialization/InitSystemorphToDatabase) notebook. Otherwise you will get a correctly structured, but completely empty DB "
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:101]: typo autmatically

Comment (andrey-katz-systemorph) [reply ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb]: fixed

File: ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                "\n    );"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "var allTypesFiltered = allTypes",
```
[the remark this answers is not in the document]
Comment (andrey-katz-systemorph) [reply ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb:44]: Actually not. I run it with the allTypes and it fails miserably. The error is:

Compilation Error (2,26): error CS1503: Argument 1: cannot convert from 'System.Collections.Generic.IEnumerable<System.Type>' to 'System.Type[]'

I am not sure, scaffolding of each class causes this failure, but it does. So let's leave it as it is.

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                "# Brief Explanation of Migration"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "Migration is a powerful tool that allows us to keep the same Database (DB) during the whole lifetime of the project, without any need to remove it if one needs to modify the DB, e.g. via adding a removing certain tables or columns. With the migration, if modification of the DB is needed, we do not remove the old DB, but rather migrate it incrementally. Note, however, that running a project with migration is a choice rather than a default option. If you are OK with removing the DB each time when some change in the tables structure is needed, migration is not necessary. ",
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:30]: "adding a removing" => "addition or removing"

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                "\n",
+                "\nAt the first step we perform the [scaffolding](Scaffold.) The scaffolding procedure naturally outputs the migration code and the model snapshot. The migration code is the code that we should run in order to change the DB according to our new requirements, or, alternatively, if we start from scratch, to create the DB. The snapshot contains all the information about the structure of the DB after the migration. It is necessary to have the snapshot, in order for us not to start from scratch at the next migration, but rather only *implement the changes*. Each time when the DB has to be changed, we should scaffold the new model, taking into the account the snapshot of the previous step. The snapshot of the new scaffold should be copied manually without changes into the [Snapshot notebook](Snapshot). We have done it alrerady for the initial migration, please do not forget to change the [Snapshot notebook](Snapshot) for any subsequent migration.",
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:32]: taking into the account => taking into account

Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:32]: alrerady  => already

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "Since the logic migration includes incremental changes, while adding or removing something from the DB, one should not delete the schema anymore. Therefore, the [schema deletion notebook](../SchemaDelete) must not be used. ",
+                "\n",
+                "\nSince migration is not a default mode, we are still leaving this notebook in the folder. However, note that now using it would contradict the logic of migration, so please do not do it light-headedly. Stick to migration if this iy your choice."
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:65]: iy => is

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                "## Configure"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "Before you run the [Configure notebook](../Configure), please uncomment the row, that specifies that the Configuration is run with migration. "
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:83]: run => execute

File: ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            "name": "C#"
+        }
+    },
+    "nbformat": 4,
+    "nbformat_minor": 5,
+    "cells": [
+        {
+            "cell_type": "markdown",
+            "source": [
+                "# Brief Explanation of Migration"
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/ReadMe.ipynb:21]: Start the NB with a big Title. You can copy the code for the markdown from other NBs

File: ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            "file_extension": ".cs",
+            "mimetype": "text/plain",
+            "name": "C#"
+        }
+    },
+    "nbformat": 4,
+    "nbformat_minor": 5,
+    "cells": [
+        {
+            "cell_type": "code",
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb:19]: Start the NB with a title (Name of the Notebook)

File: ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                "\n    .Scaffold(\"InitialTypes\");"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "In order to perform any subsequent change in the DB, there is no need to erase the DB and create from the scratch. It is sufficient to erase the cell above and uncomment the following cell. Please do not forget to insert a proper array of classes that you'd like to scaffold in the third row, as well as to give your own unique name to this migration in the 5th line."
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb:65]: and create from the scratch => and create a new from the scratch

Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/Scaffold.ipynb:65]: Can you give some description of how to get the array of classes? 
are those that have changes wrt initial?
what are the implications of the migration name?

File: ifrs17-template/Database/MigrationAndScaffolding/Snapshot.ipynb [hunk 10]
```diff
@@ -0,0 +1,1071 @@
...
+            "name": "C#"
+        }
+    },
+    "nbformat": 4,
+    "nbformat_minor": 5,
+    "cells": [
+        {
+            "cell_type": "markdown",
+            "source": [
+                "Each time you produce a new migration, please copy and paste the snapshot code directly here. This must be the copy of the *most recent* migration snapshot. "
```
Comment (dcolleoni) [inline ifrs17-template/Database/MigrationAndScaffolding/Snapshot.ipynb:21]: Add a Title

**Review by dcolleoni — APPROVED** [review]

**Pull request merged**